src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStorage.java

```diff
@@ -0,0 +1,169 @@
+     * @param lastMilestoneBeforeCompletion the last milestone the cleared run reached before completion
+     * @param milestones the currently known milestones for other running builds of the same job.
+     */
+    public record ClearResult(Integer lastMilestoneBeforeCompletion, Map<Integer, Integer> milestones) {}
+
+    /**
+     * Records a completed build.
+     * <p>
+     * This is expected to cancel all prior builds that are still ongoing.
+     *
+     * @param run the active run
+     */
+    public final void complete(@NonNull Run<?,?> run) {
+        var result = clear(run);
+        LOGGER.finest(() -> "milestones after completion: " + result.milestones);
+        if (result.lastMilestoneBeforeCompletion != null) {
+            LOGGER.finest(() -> "Build" + run + " last milestone before completion: " + result.lastMilestoneBeforeCompletion);
+            var buildsToCancel = getBuildsToCancel(run.getNumber(), Integer.MAX_VALUE, result.milestones);
+            cancel(buildsToCancel, run);
+        } else {
+            LOGGER.finest(() -> "Build " + run + " was not using milestones, nothing to cancel");
+        }
+    }
+
+    /**
+     * @return the active implementation
+     */
+    public static MilestoneStorage get() {
+        return ExtensionList.lookupFirst(MilestoneStorage.class);
+    }
+
+    /**
+     * Records a starting run.
+     *
+     * @param run the active run
+     */
+    public final void start(@NonNull Run<?, ?> run) {
+        store(run, null);
+    }
+
+    /**
+     * Called when a run resumes.
+     * @param run the active run
+     * @param ordinal the highest ordinal found for this run while resuming it.
+     */
+    public final void resume(@NonNull Run<?,?> run, @CheckForNull Integer ordinal) {
+        store(run, ordinal);
+    }
+
+    /**
+     * Cancels the given run due to another run passing a milestone.
+     * @param run the run to cancel
+     * @param externalizableId the externalizable id for the run causing the cancellation.
+     */
+    public final void cancel(Run<?, ?> run, String externalizableId) {
+        LOGGER.fine(() -> "Cancelling " + run);
+        Executor e = run.getExecutor();
+        if (e != null) {
+            e.interrupt(Result.NOT_BUILT, new CancelledCause(externalizableId));
+        } else {
+            LOGGER.warning(() -> "could not cancel an older flow because it has no assigned executor");
+        }
+    }
+
+    /**
+     * Given a reference run, cancels all the given builds of the same job with the given build numbers.
+     * @param buildNumbers The build numbers of runs to cancel.
+     * @param referenceRun The run that is causing the cancellation.
+     */
+    protected abstract void cancel(Set<Integer> buildNumbers, Run<?, ?> referenceRun);
+
+    /**
+     * The default implementation, attaching an {@link Action} to the {@link Job}.
+     */
+    @Extension(ordinal = -1)
+    public static class DefaultMilestoneStorage extends MilestoneStorage {
```

Comment:
Maybe clearer to put this inside `MilestoneJobAction` (or whatever it is), and also `@Restrict` it.